### ISSUE TITLE
'Undo and redo' Command isn't working properly 

### ISSUE THREAD

**Issue by heyBrettJosh** [opentoonz/opentoonz#4907]
### Description

The 'undo and redo' command isn't working properly, it just does the most random cutting and undoing and redoing. Here, I will list all the issues I've encountered with it:

**Undo**
**- Undoing non-recent activity.** Even though clearly the newly drawn stroked is the most recent. It undo's the old ones.
**- Cutting portion of drawings.** Supposedly its command is to undo written work, it instead cut portion of drawings in random frames which is unintentional.
**- Isn't redoing recent activity, only adds deleted parts in wrong frames and cut portions of drawing.** 
**- It's bugged, affecting other frames and while both undo and redo is issued to this, it would result for more harder saving of the project and inconvenience of shortcut techniques.**

**Redo**
**- Ends it's redo when there is obviously more to gain back.**
**- Redoing random clips/portion/cuts that's not present on that frame**
**- Mostly acts the same way as undo but its "redoing"**

Video Proof
https://github.com/opentoonz/opentoonz/assets/134044865/3d6104be-4b9c-4b5c-8e2e-1032f1f8ceab

### Steps to Reproduce

1. Open 'Opentoonz'
2. Create test projet for bug test
3. Start scribbling/stroking/crop-and-transform in each frame (Literally any animator would do)
4. Then, use Ctrl+Z to undo recent activity and Ctrl+Y to redo.
5. Issue/bug should be present like the video below

Video:
https://github.com/opentoonz/opentoonz/assets/134044865/4e390794-ab3f-47bc-8e54-b3e4afc15ee3

### Expected Behavior

After frames of animating, redo and undoing. This would be the list of expected issues to be encountered with:

**Undo**
**- Undoing non-recent activity.** Even though clearly the newly drawn stroked is the most recent. It undo's the old ones.
**- Cutting portion of drawings.** Supposedly its command is to undo written work, it instead cut portion of drawings in random frames which is unintentional.
**- Isn't redoing recent activity, only adds deleted parts in wrong frames and cut portions of drawing.** 
**- It's bugged, affecting other frames and while both undo and redo is issued to this, it would result for more harder saving of the project and inconvenience of shortcut techniques.**

**Redo**
**- Ends it's redo when there is obviously more to gain back.**
**- Redoing random clips/portion/cuts that's not present on that frame**
**- Mostly acts the same way as undo but its "redoing"**

https://github.com/opentoonz/opentoonz/assets/134044865/daa6e905-a978-4ce5-99e7-4a5db839ad45

### Screenshots, Video & Crash Logs

Video is evident enough. Should have concise screenshot in video presented.

https://github.com/opentoonz/opentoonz/assets/134044865/1c75ea02-97f4-4313-ab5f-3accd5b60ba6

### OpenToonz Version

1.7

### Operating System

Windows

### GPU

Intel

### Graphics Tablet

None

**Comment by heyBrettJosh** [comment]
Just updated the video proofs. It should have the concise evidence. Sorry for the inconvenience before.
Also yes, **selecting a portion of a drawing and moving it around**. Crop and transform doesn't really have any issues, its just part of the procedure in which will be evident enough for the bug encounter. Or you could say, its any ordinary animators job to be using crop and transform. It really is just the undo and redo bug that causes problems which happens to cut portions in different frames.

**Comment by heyBrettJosh** [comment]
Everything seems so well on your end and as well to others. I'm starting to wonder if it's the device (it's compatible even) or maybe some temporary issue that my program is running (?) I don't know how it works though but I hope this gets normal ASAP. 

Anyways, thanks for your time and consideration!

**Comment by heyBrettJosh** [comment]
Exactly. I have been refraining from using my undo and redo commands just to avoid bug collision everywhere. But now that it isn't fixed yet, it hits different when you don't have the necessary tools and commands to have consistent work considering the time management.

**Comment by RodneyBaker** [comment]
Undo shouldn't be avoided but it should be better understood.
There are certainly issues with History/Undo and this especially the case with some specific workflows (as has been mentioned here by otheres).

For the present time, Undo/Redo is best used in Opentoonz while working specifically in one area of the program.
If moving say, between drawings and between keyframing, the program will likely not be able to determine which command to undo.  Those specifica paths in the code were never put in place.

Thanks for the information you have provided as that will help as we try to make the process of history/undo/redo better in the future.

There are a number of issues related to Undo/Redo/History both open and closed in the Opentoonz github.
Those also will be useful to include in the review as we move forward.

**Comment by RodneyBaker** [comment]
Something definitely worth doing when Undo problem arise is restarting Opentoonz.
If continuing to use OT after Undo issue appear the problem will tend to just get worse.

Also, after doing a Save All (and making sure no other instance of Opentoonz is open) go to File > Clear Cache Folder to clear out the cobwebs of old file pointers.  

If not using the latest nightly release... highly recommended.
Link to latest nightlies:  https://github.com/opentoonz/opentoonz_nightlies/releases
(all nightlies are cumulative so the latest release has all prior updates)

**Comment by RodneyBaker** [comment]
BTW @heyBrettJosh Thanks much for the videos.  That helps tremendously in seeing what is going on.

Added:  From what I can tell this particular issue with Undo has been resolved in the latest nightly release.  Keeping this report open for now to collect other related issues.

**Comment by RodneyBaker** [comment]
Closing as other reports related to Undo/Redo history are still open.
For instance:  #4964
